plugins/hive-plugin/src/main/java/edu/berkeley/ground/plugins/hive/GDatabase.java

```diff
@@ -0,0 +1,362 @@
+package edu.berkeley.ground.plugins.hive;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.hadoop.hive.metastore.api.Database;
+import org.apache.hadoop.hive.metastore.api.InvalidInputException;
+import org.apache.hadoop.hive.metastore.api.InvalidObjectException;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.partition.spec.PartitionSpecProxy;
+import org.apache.hive.common.util.HiveStringUtils;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+import edu.berkeley.ground.api.models.Edge;
+import edu.berkeley.ground.api.models.EdgeVersion;
+import edu.berkeley.ground.api.models.Node;
+import edu.berkeley.ground.api.models.NodeFactory;
+import edu.berkeley.ground.api.models.NodeVersion;
+import edu.berkeley.ground.api.models.NodeVersionFactory;
+import edu.berkeley.ground.api.models.Structure;
+import edu.berkeley.ground.api.models.StructureVersion;
+import edu.berkeley.ground.api.models.Tag;
+import edu.berkeley.ground.api.versions.GroundType;
+import edu.berkeley.ground.exceptions.GroundException;
+import edu.berkeley.ground.plugins.hive.GroundStore.EntityState;
+
+public class GDatabase {
+    static final private Logger LOG = LoggerFactory.getLogger(GDatabase.class.getName());
+
+    static final String DATABASE_NODE = "_DATABASE";
+
+    static final String DATABASE_TABLE_EDGE = "_DATABASE_TABLE";
+
+    private static final List<String> EMPTY_PARENT_LIST = new ArrayList<String>();
+
+    private GroundReadWrite ground = null;
+    private GTable table = null;
+
+    public GDatabase(GroundReadWrite ground) {
+        this.ground = ground;
+        this.table = new GTable(ground);
+    }
+
+    public Node getNode(String dbName) throws GroundException {
+        try {
+            LOG.debug("Fetching database node: " + dbName);
+            return ground.getNodeFactory().retrieveFromDatabase(dbName);
+        } catch (GroundException ge1) {
+            LOG.debug("Not found - Creating databsae node: " + dbName);
+
+            Node node = ground.getNodeFactory().create(dbName);
+            Structure nodeStruct = ground.getStructureFactory().create(node.getName());
+
+            return node;
+        }
+    }
+
+    public Structure getNodeStructure(String dbName) throws GroundException {
+        try {
```

Comment:
What's happening in this method? What is a `NodeStructure`? and why is the `Node` being retrieved?